x-pack/legacy/plugins/siem/public/components/embeddables/map_tool_tip/map_tool_tip.tsx

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useEffect, useState } from 'react';
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiLoadingSpinner,
+  EuiOutsideClickDetector,
+} from '@elastic/eui';
+import { FeatureGeometry, FeatureProperty, MapToolTipProps } from '../types';
+import { DraggablePortalContext } from '../../drag_and_drop/draggable_wrapper';
+import { ToolTipFooter } from './tooltip_footer';
+import { LineToolTipContent } from './line_tool_tip_content';
+import { PointToolTipContent } from './point_tool_tip_content';
+import { Loader } from '../../loader';
+
+export const MapToolTip = React.memo<MapToolTipProps>(
+  ({
+    addFilters,
+    closeTooltip,
+    features = [],
+    isLocked,
+    getLayerName,
+    loadFeatureProperties,
+    loadFeatureGeometry,
+  }) => {
+    const [isLoading, setIsLoading] = useState<boolean>(true);
+    const [isLoadingNextFeature, setIsLoadingNextFeature] = useState<boolean>(false);
+    const [featureIndex, setFeatureIndex] = useState<number>(0);
+    const [featureProps, setFeatureProps] = useState<FeatureProperty[]>([]);
+    const [featurePropsFilters, setFeaturePropsFilters] = useState<Record<string, object>>({});
+    const [featureGeometry, setFeatureGeometry] = useState<FeatureGeometry | null>(null);
+    const [, setLayerName] = useState<string>('');
+
+    useEffect(() => {
+      // Early return if component doesn't yet have props -- result of mounting in portal before actual rendering
+      if (
+        features.length === 0 ||
+        getLayerName == null ||
+        loadFeatureProperties == null ||
+        loadFeatureGeometry == null
+      ) {
+        return;
+      }
+
+      // Separate loaders for initial load vs loading next feature to keep tooltip from drastically resizing
+      if (!isLoadingNextFeature) {
+        setIsLoading(true);
+      }
+
+      const fetchFeatureProps = async () => {
+        if (features[featureIndex] != null) {
+          const [featureProperties, featureGeo, layerNameString] = await Promise.all([
+            loadFeatureProperties({
+              layerId: features[featureIndex].layerId,
+              featureId: features[featureIndex].id,
+            }),
+            loadFeatureGeometry({
+              layerId: features[featureIndex].layerId,
+              featureId: features[featureIndex].id,
+            }),
+            getLayerName(features[featureIndex].layerId),
```

Comment:
optional nit: replace all the repeated `features[featureIndex].layerId` and `features[featureIndex].id` with

```
const layerId = features[featureIndex].layerId;
const featureId = features[featureIndex].id;
```

and then just refer to `layerId`, `featureId`